From 33dcb5faf2cee31fe20def08574ea3d98b980a18 Mon Sep 17 00:00:00 2001
From: Sven van Haastregt <sven.vanhaastregt@arm.com>
Date: Thu, 9 Jun 2022 09:38:02 +0100
Subject: [PATCH 046/172] [NFC] Use only one PassManager in SPIRVWriter

Factor out common code, in preparation of transitioning SPIRVWriter to
the new PassManager.

(cherry picked from commit dc9e645917b106332cdf46a790a9481d44282a96)
---
 lib/SPIRV/SPIRVWriter.cpp | 58 +++++++++++++++++++++++----------------
 1 file changed, 35 insertions(+), 23 deletions(-)

diff --git a/lib/SPIRV/SPIRVWriter.cpp b/lib/SPIRV/SPIRVWriter.cpp
index f8d4cbb4..b290f426 100644
--- a/lib/SPIRV/SPIRVWriter.cpp
+++ b/lib/SPIRV/SPIRVWriter.cpp
@@ -5066,35 +5066,54 @@ bool isValidLLVMModule(Module *M, SPIRVErrorLog &ErrorLog) {
   return true;
 }
 
-bool llvm::writeSpirv(Module *M, std::ostream &OS, std::string &ErrMsg) {
-  SPIRV::TranslatorOpts DefaultOpts;
-  // To preserve old behavior of the translator, let's enable all extensions
-  // by default in this API
-  DefaultOpts.enableAllExtensions();
-  return llvm::writeSpirv(M, DefaultOpts, OS, ErrMsg);
-}
+namespace {
+
+bool runSpirvWriterPasses(Module *M, std::ostream *OS, std::string &ErrMsg,
+                          const SPIRV::TranslatorOpts &Opts) {
+  // Perform the conversion and write the resulting SPIR-V if an ostream has
+  // been given; otherwise only perform regularization.
+  bool WriteSpirv = OS != nullptr;
 
-bool llvm::writeSpirv(Module *M, const SPIRV::TranslatorOpts &Opts,
-                      std::ostream &OS, std::string &ErrMsg) {
   std::unique_ptr<SPIRVModule> BM(SPIRVModule::createSPIRVModule(Opts));
   if (!isValidLLVMModule(M, BM->getErrorLog()))
     return false;
 
   legacy::PassManager PassMgr;
   addPassesForSPIRV(PassMgr, Opts);
-  // Run loop simplify pass in order to avoid duplicate OpLoopMerge
-  // instruction. It can happen in case of continue operand in the loop.
-  if (hasLoopMetadata(M))
-    PassMgr.add(createLoopSimplifyPass());
-  PassMgr.add(createLLVMToSPIRVLegacy(BM.get()));
+  if (WriteSpirv) {
+    // Run loop simplify pass in order to avoid duplicate OpLoopMerge
+    // instruction. It can happen in case of continue operand in the loop.
+    if (hasLoopMetadata(M))
+      PassMgr.add(createLoopSimplifyPass());
+    PassMgr.add(createLLVMToSPIRVLegacy(BM.get()));
+  }
+
   PassMgr.run(*M);
 
   if (BM->getError(ErrMsg) != SPIRVEC_Success)
     return false;
-  OS << *BM;
+
+  if (WriteSpirv)
+    *OS << *BM;
+
   return true;
 }
 
+} // namespace
+
+bool llvm::writeSpirv(Module *M, std::ostream &OS, std::string &ErrMsg) {
+  SPIRV::TranslatorOpts DefaultOpts;
+  // To preserve old behavior of the translator, let's enable all extensions
+  // by default in this API
+  DefaultOpts.enableAllExtensions();
+  return llvm::writeSpirv(M, DefaultOpts, OS, ErrMsg);
+}
+
+bool llvm::writeSpirv(Module *M, const SPIRV::TranslatorOpts &Opts,
+                      std::ostream &OS, std::string &ErrMsg) {
+  return runSpirvWriterPasses(M, &OS, ErrMsg, Opts);
+}
+
 bool llvm::regularizeLlvmForSpirv(Module *M, std::string &ErrMsg) {
   SPIRV::TranslatorOpts DefaultOpts;
   // To preserve old behavior of the translator, let's enable all extensions
@@ -5105,12 +5124,5 @@ bool llvm::regularizeLlvmForSpirv(Module *M, std::string &ErrMsg) {
 
 bool llvm::regularizeLlvmForSpirv(Module *M, std::string &ErrMsg,
                                   const SPIRV::TranslatorOpts &Opts) {
-  std::unique_ptr<SPIRVModule> BM(SPIRVModule::createSPIRVModule());
-  if (!isValidLLVMModule(M, BM->getErrorLog()))
-    return false;
-
-  legacy::PassManager PassMgr;
-  addPassesForSPIRV(PassMgr, Opts);
-  PassMgr.run(*M);
-  return true;
+  return runSpirvWriterPasses(M, nullptr, ErrMsg, Opts);
 }
-- 
2.20.1

